modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsRepositoryImpl.java

```diff
@@ -0,0 +1,273 @@
+/*
+ * Copyright 2020 GridGain Systems, Inc. and Contributors.
+ *
+ * Licensed under the GridGain Community Edition License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.gridgain.com/products/software/community-edition/gridgain-community-edition-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.ignite.internal.processors.query.stat;
+
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.internal.GridKernalContext;
+import org.apache.ignite.internal.processors.cache.GridCacheUtils;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Statistics repository implementation.
+ */
+public class IgniteStatisticsRepositoryImpl implements IgniteStatisticsRepository {
+    /** Logger. */
+    private IgniteLogger log;
+
+    /** Kernal context. */
+    private final GridKernalContext ctx;
+
+    /** Table->Partition->Partition Statistics map, populated only on server nodes without persistence enabled. */
+    private final Map<StatsKey, Map<Integer, ObjectPartitionStatisticsImpl>> partsStats;
+
+    /** Local (for current node) object statistics. */
+    private final Map<StatsKey, ObjectStatisticsImpl> localStats;
+
+    /** Global (for whole cluster) object statistics. */
+    private final Map<StatsKey, ObjectStatisticsImpl> globalStats = new ConcurrentHashMap<>();
+
+    /**
+     * Constructor.
+     *
+     * @param ctx Kernal context.
+     */
+    public IgniteStatisticsRepositoryImpl(GridKernalContext ctx) {
+        this.ctx = ctx;
+
+        if (ctx.config().isClientMode() || ctx.isDaemon()) {
+            // Cache only global statistics, no store
+            partsStats = null;
+            localStats = null;
+        } else {
+            // Persistence store
+            partsStats = GridCacheUtils.isPersistenceEnabled(ctx.config()) ? null : new ConcurrentHashMap<>();
+
+            localStats = new ConcurrentHashMap<>();
+        }
+        log = ctx.log(IgniteStatisticsRepositoryImpl.class);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void saveLocalPartitionsStatistics(StatsKey key,
+                                                        Collection<ObjectPartitionStatisticsImpl> statistics,
+                                                        boolean fullStat) {
+        if (partsStats != null) {
```

Comment:
is it a valid case where someone tries to save partition statistics on client node?

also let's invert condition